### PR TITLE
Reduce coupling in Population and GA now takes initial solutions

### LINKED ISSUE #71: Construct Population with an initial pool of solutions
Rather than having the Population initialise its own individuals.

### PR DESCRIPTION
**Pull request by leonlan**
This PR:
- Population no longer depends on problem data, penalty manager and Xorshift.
- Closes #71: initial population solutions are no longer initialised in Population itself.
- GeneticAlgorithm now takes as `initial_solutions` argument, which is required to be non-empty.

### CODE REVIEW

**Comment by leonlan** [comment]
@N-Wouda This is a proposal to close #71, to address the discussion in #156, and to address coupling in Population #157. I have not yet changed tests and notebooks, which I'll do if we decide to proceed with these changes.

File: pyvrp/Population.py
Comment (leonlan) [inline pyvrp/Population.py:34]: We can also remove the RNG dependency and pass it as argument to `self.select`/`self.binary_tournament`. I don't know if that's desirable.

File: pyvrp/Population.py
Comment (N-Wouda) [inline pyvrp/Population.py:142]: Do you know if this impacts performance? I don't really have a good intuition for this.

File: pyvrp/Population.py
Comment (N-Wouda) [reply pyvrp/Population.py:34]: I think it's better because there's less internal state on `Population` then.

File: pyvrp/Population.py
Comment (N-Wouda) [reply pyvrp/Population.py:142]: I'd assume no, but it might not hurt to make sure.

Comment (leonlan) [reply pyvrp/Population.py:143]: I'll run a benchmark just to be sure.

**Comment by leonlan** [comment]
Benchmark: GH1000, 3600s, 1 seed

[3f14e09](https://github.com/N-Wouda/PyVRP/pull/168/commits/3f14e09a63bf8cd7cfcc5070e344a3bba8ac6b1c)
```
Cost:  333780.9
Iters: 102301.9
```

main
```
TODO
```

**Comment by wouterkool** [comment]
Shouldn't we use a different set of initial solutions when restarting?

**Comment by leonlan** [comment]
> Shouldn't we use a different set of initial solutions when restarting?

I don't think it matters. The binary selection, crossovers, and local search introduce randomness. So even if we restart with the same initial population, the resulting trajectory will be different each time.

**Comment by leonlan** [comment]
This PR is ready to be reviewed. Meanwhile, I'll run a 10 seed benchmark to check that there is no performance drop with the new change (restart with the same individuals).

**Comment by N-Wouda** [comment]
I'll try to have a look at this tonight.

File: pyvrp/Population.py
Comment (N-Wouda) [inline pyvrp/Population.py:138]: Typo: agains.

Can we rewrite this a bit to really clarify that we're restarting from the initial individuals that were passed when this object was initialised? It's not clear enough to me what "original initial individuals" means.

File: pyvrp/Population.py
Comment (N-Wouda) [inline pyvrp/Population.py:17]:
```suggestion
        List of individuals used to initialise the population.
```

File: pyvrp/Population.py
Comment (N-Wouda) [inline pyvrp/Population.py:39]:
```suggestion
        for individual in initial_solutions:
            self.add(individual)
```

File: pyvrp/Population.py
Comment (N-Wouda) [inline pyvrp/Population.py:144]:
```suggestion
        for individual in self._initial_solutions:
            self.add(individual)
```

File: pyvrp/plotting/tests/test_plotting.py
Comment (N-Wouda) [inline pyvrp/plotting/tests/test_plotting.py:69]: The spacing here doesn't make much sense to me. I'd expect `params`, `init`, and `pop` to all be together in the same paragraph (since they're all part of "population stuff").

File: pyvrp/tests/test_Population.py
Comment (N-Wouda) [inline pyvrp/tests/test_Population.py:280]: Why does this need a different (new) instance?

File: pyvrp/tests/data/RC208LessVehicles.txt
Comment (N-Wouda) [inline pyvrp/tests/data/RC208LessVehicles.txt:1]: If we can, I'd like to avoid introducing new test data.

File: pyvrp/tests/test_Population.py
Comment (N-Wouda) [inline pyvrp/tests/test_Population.py:200]: I don't really get why this needs to be changed. Would something like:
```python
init = set(...)
current = {individual for individual in pop}
assert_(len(current & init), params.min_pop_size)

pop.restart()

new = {individual for individual in pop}
assert_(len(new & init), params.min_pop_size)
```
not work?

File: pyvrp/tests/test_Population.py
Comment (N-Wouda) [inline pyvrp/tests/test_Population.py:205]: Do we need a new test data instance for this?

File: pyvrp/tests/helpers.py
Comment (N-Wouda) [inline pyvrp/tests/helpers.py:45]: Can we have the order of the arguments here be the same as they are for almost every other thing: `data, pm, rng`? I'm fine with `num_sols` coming first.

File: pyvrp/Population.py
Comment (N-Wouda) [inline pyvrp/Population.py:27]: Can we make the type here a general iterable? It doesn't necessarily have to be a list.

File: pyvrp/tests/data/OkSmallFeasible.txt
Comment (N-Wouda) [inline pyvrp/tests/data/OkSmallFeasible.txt:1]: If we can, I'd like to avoid introducing new test data.

File: pyvrp/tests/helpers.py
Comment (N-Wouda) [inline pyvrp/tests/helpers.py:43]:
```suggestion
    Returns a list of ``num_sols`` random solutions.
```

File: pyvrp/tests/test_Population.py
Comment (N-Wouda) [inline pyvrp/tests/test_Population.py:217]: This isn't true in general. Suppse we add `15` individuals. Then `15 % 11 == 4`. But what if we've added 7 feasible and 8 infeasible solutions? Then no purge would have been triggered because neither subpopulation would have been of maximal size.

The reason this test passes is that all solutions are feasible in this instance, which is not explicitly mentioned in the test. Can we document that much more explicitly?

File: pyvrp/Population.py
Comment (N-Wouda) [reply pyvrp/Population.py:27]: I think I'd make sense for `initial_solutions` to be optional. By default we construct an empty population. This'd be possible if we did, for example,
```python
def __init__(
  self,
  diversity_op,
  initial_solutions: Iterable[Individual] = tuple(),
  params: PopulationParams = PopulationParams()
)
```

File: pyvrp/tests/test_GeneticAlgorithm.py
Comment (N-Wouda) [inline pyvrp/tests/test_GeneticAlgorithm.py:105]: While we're here anyway:

```suggestion
    ls = LocalSearch(data, pm, rng, compute_neighbours(data))
    ls.add_node_operator(Exchange10(data, pm))
```

File: pyvrp/tests/test_Population.py
Comment (N-Wouda) [inline pyvrp/tests/test_Population.py:147]: I was thinking about whether we should make `params` a keyword only argument. Perhaps yes, but then not in this PR (since we should then probably enforce it all at once for every Params-like object).

**Review by N-Wouda — COMMENTED** [review]
A bunch of small things and some questions. On the whole this looks good to me!

File: pyvrp/tests/test_Population.py
Comment (leonlan) [reply pyvrp/tests/test_Population.py:280]: Because the test assumes that the randomly generated individuals are all infeasible, which was not the case when I ran the test with the original file.

File: pyvrp/Population.py
Comment (leonlan) [reply pyvrp/Population.py:27]: I don't think that making the initial solution argument is desirable. Creating an empty Population is not a good default, and it also doesn't work well with the restarting mechanism.

File: pyvrp/tests/test_Population.py
Comment (N-Wouda) [reply pyvrp/tests/test_Population.py:280]: But we can just filter the feasible solutions away before passing `init` to `pop`, right?

File: pyvrp/Population.py
Comment (N-Wouda) [reply pyvrp/Population.py:27]: >Creating an empty Population is not a good default

Why not? That Population should not be empty is a problem for the GeneticAlgorithm, not for the Population container itself.

>and it also doesn't work well with the restarting mechanism.

This I don't understand. Restart would just return to an empty population again; I don't see what'd go wrong.

Comment (N-Wouda) [reply pyvrp/Population.py:27]: Thinking about this, we should probably enforce in the GA that the population has least 1/2 (I think 2?) individuals, right now it can be empty.

Comment (N-Wouda) [reply pyvrp/Population.py:27]: I'll add a PR for that.

File: pyvrp/tests/test_Population.py
Comment (leonlan) [reply pyvrp/tests/test_Population.py:200]: No idea why I changed this - I changed it again to what you suggested!

**Comment by leonlan** [comment]
I'll address the remaining comments tomorrow morning.

File: pyvrp/GeneticAlgorithm.py
Comment (N-Wouda) [inline pyvrp/GeneticAlgorithm.py:98]: Here, and in a test: this is a bit too clever IMHO. I had to think for too long about why this works, so its readability could be improved.

Something like
```python
# Find best feasible initial solution if any exist, else set lower cost
# solution as initial best.
init_feas = [indiv for indiv in initial_solutions if indiv.is_feasible()]
sols = init_feas if init_feas else initial_solutions
self._best = min(sols, key=lambda indiv: indiv.cost())
```
is barely longer, but doesn't make me think as hard.

In the test I think we can add some feasible individual that we know is the best, and then test whether that individual is indeed set as the best. There's no need to reproduce the logic there.

**Review by N-Wouda — APPROVED** [review]
LGTM! A few small things about how these changes can be made a bit more legible, but nothing that's blocking merging this in.

**Comment by leonlan** [comment]
> LGTM! A few small things about how these changes can be made a bit more legible, but nothing that's blocking merging this in.

OK, I addressed all comments. Thanks for all the help! I'll run a benchmark before merging this PR.

**Comment by leonlan** [comment]
GH1000, 10 seeds, 36000s
```
min     33341.5
mean    33359.6
max     33374.5
```
All looks fine, so I'm merging this in.

**Pull request merged**
